### PR TITLE
Reinstall syncer when in pull mode and image changes

### PR DESCRIPTION
**Pull request by eddycharly**
This PR makes the cluster controller reinstall syncer when in pull mode and syncer image has changed.

### CODE REVIEW

File: pkg/reconciler/cluster/syncer.go [hunk 1]
```diff
@@ -267,3 +267,20 @@ func healthcheckSyncer(ctx context.Context, client kubernetes.Interface, logical
 	}
 	return nil
 }
+
+func isSyncerInstalledAndUpToDate(ctx context.Context, client kubernetes.Interface, logicalCluster, syncerImage string) (bool, error) {
+	pods, err := client.CoreV1().Pods(syncerNS).List(ctx, metav1.ListOptions{LabelSelector: "app=" + syncerWorkloadName(logicalCluster)})
```
[the remark this answers is not in the document]
Comment (eddycharly) [reply pkg/reconciler/cluster/syncer.go]: ✅

File: pkg/reconciler/cluster/syncer.go [hunk 2]
```diff
@@ -267,3 +267,20 @@ func healthcheckSyncer(ctx context.Context, client kubernetes.Interface, logical
...
+	if err != nil {
+		return false, err
+	}
+	if len(pods.Items) == 0 {
+		return false, nil
+	}
+	if len(pods.Items) > 1 {
+		return true, fmt.Errorf("Syncer pod not ready: there should be only 1 syncer pod")
+	}
+	if pods.Items[0].Spec.Containers[0].Image != syncerImage {
```
[the remark this answers is not in the document]
Comment (eddycharly) [reply pkg/reconciler/cluster/syncer.go]: ✅

File: pkg/reconciler/cluster/cluster.go [hunk 3]
```diff
@@ -107,7 +107,19 @@ func (c *Controller) reconcile(ctx context.Context, cluster *clusterv1alpha1.Clu
...
 
-	if !sets.NewString(cluster.Status.SyncedResources...).Equal(groupResources) {
+	syncerInstallNeeded := false
+	if c.syncerMode == SyncerModePull {
+		upToDate, err := isSyncerInstalledAndUpToDate(ctx, client, logicalCluster, c.syncerImage)
+		if err != nil {
+			klog.Errorf("error checking if syncer needs to be installed: %v", err)
+			return err
+		}
+		if !upToDate {
```
Comment (stevekuznetsov) [inline pkg/reconciler/cluster/cluster.go:117]: Is there any reason to have two variables (opposed in meaning) for one check? We could e.g.:


```go
var upToDate bool
if c.syncerMode == SyncerModePull {
    var err error 
    upToDate, err = isSyncerInstalledAndUpToDate(...
}

if !sets... || !upToDate {

}
```

Comment (eddycharly) [reply pkg/reconciler/cluster/cluster.go]: Fixed.
It looks like there is room for improvement in the install/uninstall process (for example, when going from pull mode to push mode, syncer should be uninstalled). I'll address this in another PR.

Comment (eddycharly) [reply pkg/reconciler/cluster/cluster.go]: I opened https://github.com/kcp-dev/kcp/pull/357 to improve `syncer` install/uninstall process.

**Review by stevekuznetsov — APPROVED** [review]

**Pull request merged**